tests/api_tests/nodes/test_views.py

```diff
@@ -879,6 +879,514 @@ def test_filtering_node_with_non_bibliographic_contributor(self):
+            'permission': 'write',
+            'bibliographic': True
+        }
+        res = self.app.post_json(self.private_url, data, auth=self.user.auth)
+        assert_equal(res.status_code, 201)
+        assert_equal(res.json['data']['id'], self.user_two._id)
+
+        self.private_project.reload()
+        assert_in(self.user_two, self.private_project.contributors)
+        assert_equal(self.private_project.get_permissions(self.user_two), ['read', 'write'])
+
+    def test_adds_read_contributor_private_project_admin(self):
+        data = {
+            'id': self.user_two._id,
+            'permission': 'read',
+            'bibliographic': True
+        }
+        res = self.app.post_json(self.private_url, data, auth=self.user.auth)
+        assert_equal(res.status_code, 201)
+        assert_equal(res.json['data']['id'], self.user_two._id)
+
+        self.private_project.reload()
+        assert_in(self.user_two, self.private_project.contributors)
+        assert_equal(self.private_project.get_permissions(self.user_two), ['read'])
+
+    def test_adds_invalid_permission_contributor_private_project_admin(self):
+        data = {
+            'id': self.user_two._id,
+            'permission': 'invalid',
+            'bibliographic': True
+        }
+        res = self.app.post_json(self.private_url, data, auth=self.user.auth, expect_errors=True)
+        assert_equal(res.status_code, 400)
+
+        self.private_project.reload()
+        assert_not_in(self.user_two, self.private_project.contributors)
+
+    def test_adds_none_permission_contributor_private_project_admin(self):
+        data = {
+            'id': self.user_two._id,
+            'permission': None,
+            'bibliographic': True
+        }
+        res = self.app.post_json(self.private_url, data, auth=self.user.auth, expect_errors=True)
+        assert_equal(res.status_code, 201)
+
+        self.private_project.reload()
+        assert_in(self.user_two, self.private_project.contributors)
+
+    def test_adds_already_existing_contributor_private_project_admin(self):
+        self.private_project.add_contributor(self.user_two, auth=Auth(self.user), save=True)
+        self.private_project.reload()
+
+        res = self.app.post_json(self.private_url, self.data_user_two,
+                                 auth=self.user.auth, expect_errors=True)
+        assert_equal(res.status_code, 400)
+
+    def test_adds_non_existing_user_private_project_admin(self):
+        data = {
+            'id': 'Fake',
+            'bibliographic': True
+        }
+        res = self.app.post_json(self.private_url, data, auth=self.user.auth, expect_errors=True)
+        assert_equal(res.status_code, 404)
+
+        self.private_project.reload()
+        assert_equal(len(self.private_project.contributors), 1)
+
+    def test_adds_contributor_private_project_non_admin(self):
+        self.private_project.add_contributor(self.user_two, permissions=['read', 'write'], auth=Auth(self.user))
```

Comment:
The model is not saved here; check that this test is failing for the expected reason. 

(It may be failing because the user has *no* authorization on this project, rather than just because they are not an admin)